providers/rpc/rpc.go

```diff
@@ -0,0 +1,351 @@
+	switch strings.ToLower(state) {
+	case "on", "off", "cycle":
+		p := RequestPayload{
+			ID:     int64(time.Now().UnixNano()),
+			Host:   c.Host,
+			Method: PowerSetMethod,
+			Params: PowerSetParams{
+				State: strings.ToLower(state),
+			},
+		}
+		resp, err := c.process(ctx, p)
+		if err != nil {
+			return ok, err
+		}
+		if resp.Error != nil {
+			return ok, fmt.Errorf("error from rpc consumer: %v", resp.Error)
+		}
+
+		return true, nil
+	}
+
+	return false, errors.New("requested power state is not supported")
+}
+
+// PowerStateGet gets the power state of a BMC machine.
+func (c *Config) PowerStateGet(ctx context.Context) (state string, err error) {
+	p := RequestPayload{
+		ID:     int64(time.Now().UnixNano()),
+		Host:   c.Host,
+		Method: PowerGetMethod,
+	}
+	resp, err := c.process(ctx, p)
+	if err != nil {
+		return "", err
+	}
+	if resp.Error != nil {
+		return "", fmt.Errorf("error from rpc consumer: %v", resp.Error)
+	}
+
+	return resp.Result.(string), nil
+}
+
+// process is the main function for the roundtrip of rpc calls to the ConsumerURL.
+func (c *Config) process(ctx context.Context, p RequestPayload) (ResponsePayload, error) {
+	// 1. create the HTTP request.
+	// 2. create the signature payload.
+	// 3. sign the signature payload.
+	// 4. add signatures to the request as headers.
+	// 5. request/response round trip.
+	// 6. handle the response.
+	req, err := c.createRequest(ctx, p)
+	if err != nil {
+		return ResponsePayload{}, err
+	}
+
+	// create the signature payload
+	// get the body and reset it as readers can only be read once.
+	body, err := io.ReadAll(req.Body)
+	if err != nil {
+		return ResponsePayload{}, err
+	}
+	req.Body = io.NopCloser(bytes.NewBuffer(body))
+	headersForSig := http.Header{}
+	for _, h := range c.Opts.Signature.IncludedPayloadHeaders {
+		if val := req.Header.Get(h); val != "" {
+			headersForSig.Add(h, val)
+		}
+	}
+	sigPay := createSignaturePayload(body, headersForSig)
+
+	// sign the signature payload
+	sigs, err := sign(sigPay, c.Opts.HMAC.Hashes, c.Opts.HMAC.PrefixSigDisabled)
+	if err != nil {
+		return ResponsePayload{}, err
+	}
+
+	// add signatures to the request as headers.
+	for algo, keys := range sigs {
+		if len(sigs) > 0 {
+			h := c.Opts.Signature.HeaderName
+			if !c.Opts.Signature.AppendAlgoToHeaderDisabled {
+				h = fmt.Sprintf("%s-%s", h, algo.ToShort())
+			}
+			req.Header.Add(h, strings.Join(keys, ","))
+		}
+	}
+
+	// request/response round trip.
+	kvs := requestKVS(req)
+	kvs = append(kvs, []interface{}{"host", c.Host, "method", p.Method, "consumerURL", c.ConsumerURL}...)
+	if p.Params != nil {
+		kvs = append(kvs, []interface{}{"params", p.Params}...)
+	}
+
+	resp, err := c.Opts.Request.Client.Do(req)
+	if err != nil {
+		c.Logger.Error(err, "failed to send rpc notification", kvs...)
+		return ResponsePayload{}, err
+	}
+	defer resp.Body.Close()
+
+	// handle the response
+	rp, err := c.handleResponse(resp, kvs)
```

Comment:
Im guessing at this point the consumer is trusted, how is this trust established?